spec/data-types-and-values.html

```diff
@@ -8,11 +8,50 @@ <h1>ECMAScript Language Types</h1>
     <emu-clause id="sec-ecmascript-language-types-record-type">
       <h1>The Record Type</h1>
       <p>TBD</p>
+
+      <emu-clause id="sec-recordtostring" aoid="RecordToString">
+        <h1>RecordToString ( _argument_ )</h1>
+        <p>The abstract operation RecordToString takes argument _argument_ (a Record) and convertes _argument_ to a String. It performs the following steps when called:</p>
+        <emu-alg>
+            1. Assert: Type(_argument_) is Record.
+            1. Return *"[record]"*.
+        </emu-alg>
+      </emu-clause>
     </emu-clause>
 
     <emu-clause id="sec-ecmascript-language-types-tuple-type">
       <h1>The Tuple Type</h1>
       <p>TBD</p>
+
+      <emu-clause id="sec-tupletostring" aoid="TupleToString">
+        <h1>TupleToString ( _argument_, [ _separator_, [ _mapper_ ] ] )</h1>
+        <p>The abstract operation TupleToString takes arguments _argument_, _separator_, and _mapper_ (a function object). _mapper_ will be invoked with *undefined* as the receiver. It performs the following steps when called:</p>
+        <emu-alg>
+            1. Assert: Type(_argument_) is Tuple.
+            1. Assert: _mapper_ is *undefined* or IsCallable(_mapper_) is *true*.
+            1. Assert: Type(_separator_) is Undefined or String.
+            1. Let _list_ be _argument_'s [[Sequence]] value.
+            1. Let _len_ be the length of List _list_.
+            1. Let _k_ be 0.
+            1. Let _R_ be the empty String.
+            1. If _separator_ is *undefined*,
+              1. Let _realSep_ be *","*.
+            1. Else,
+              1. Let _realSep_ be _separator_.
+            1. Repeat, while _k_ &lt; _len_,
+              1. If _k_ &gt; 0, then
+                1. Set _R_ to the string-concatenation of _R_ and _realSep_.
+              1. Let _item_ be _list_[_k_].
+              If _mapper_

```

Comment:
a lot of this seems like it could/should be an abstract operation, shared by both this and https://tc39.es/ecma262/#sec-array.prototype.join ?

Reply:
It definitely wouldn't be fine to look up the "join" property on Array.prototype :-) i'm suggesting it should have the same semantics via a shared abstract operation.

Reply:
Sure, but I think calling the original value %Array.prototype.join% would be safe as well and do what we want here.

Reply:
I've simplified this based on @littledan 's suggestion to just invoke %Array.prototype.join% for `TupleToString`.